src/main/java/org/kuali/kfs/kns/kim/type/DataDictionaryTypeServiceBase.java

```diff
@@ -0,0 +1,874 @@
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.kuali.kfs.core.api.uif.AttributeError;
+import org.kuali.kfs.core.api.util.type.KualiDecimal;
+import org.kuali.kfs.core.api.util.type.TypeUtils;
+import org.kuali.kfs.core.web.format.Formatter;
+import org.kuali.kfs.datadictionary.legacy.BusinessObjectDictionaryService;
+import org.kuali.kfs.datadictionary.legacy.DataDictionaryService;
+import org.kuali.kfs.kew.doctype.bo.DocumentType;
+import org.kuali.kfs.kew.doctype.service.DocumentTypeService;
+import org.kuali.kfs.kew.service.KEWServiceLocator;
+import org.kuali.kfs.kim.api.type.KimAttributeField;
+import org.kuali.kfs.kim.api.type.KimTypeInfoService;
+import org.kuali.kfs.kim.api.type.QuickFinder;
+import org.kuali.kfs.kim.bo.impl.KimAttributes;
+import org.kuali.kfs.kim.framework.type.KimTypeService;
+import org.kuali.kfs.kim.impl.type.KimType;
+import org.kuali.kfs.kim.impl.type.KimTypeAttribute;
+import org.kuali.kfs.kns.lookup.LookupUtils;
+import org.kuali.kfs.kns.service.DictionaryValidationService;
+import org.kuali.kfs.kns.service.KNSServiceLocator;
+import org.kuali.kfs.kns.util.FieldUtils;
+import org.kuali.kfs.kns.web.ui.Field;
+import org.kuali.kfs.krad.bo.BusinessObject;
+import org.kuali.kfs.krad.comparator.StringValueComparator;
+import org.kuali.kfs.krad.datadictionary.AttributeDefinition;
+import org.kuali.kfs.krad.datadictionary.PrimitiveAttributeDefinition;
+import org.kuali.kfs.krad.datadictionary.RelationshipDefinition;
+import org.kuali.kfs.krad.service.BusinessObjectService;
+import org.kuali.kfs.krad.service.KRADServiceLocator;
+import org.kuali.kfs.krad.service.KRADServiceLocatorWeb;
+import org.kuali.kfs.krad.service.ModuleService;
+import org.kuali.kfs.krad.util.ErrorMessage;
+import org.kuali.kfs.krad.util.ExternalizableBusinessObjectUtils;
+import org.kuali.kfs.krad.util.GlobalVariables;
+import org.kuali.kfs.krad.util.KRADUtils;
+import org.kuali.kfs.krad.util.ObjectUtils;
+import org.kuali.kfs.sys.KFSKeyConstants;
+import org.springframework.util.AutoPopulatingList;
+
+import java.beans.PropertyDescriptor;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * A base class for {@code KimTypeService} implementations which read attribute-related information from the Data
+ * Dictionary. This implementation is currently written against the KNS apis for Data Dictionary. Additionally, it
+ * supports the ability to read non-Data Dictionary attribute information from the {@link KimTypeInfoService}.
+ */
+public class DataDictionaryTypeServiceBase implements KimTypeService {
```

Comment:
What I meant was that you could still place your new code in an override of the `validateAttributeFormat()` method, but instead of placing it in the base class, you could place it in the class of the specific role type that uses it (`AccountingOrganizationHierarchyReviewRoleTypeServiceImpl` in this case). That should still allow it to catch the bad KIM data before it gets stored in the database.

On the other hand, given that this is only a minor code change that executes quickly and shouldn't interfere with other role types, I'm okay with leaving the customization in the base class, so that it will automatically handle any future role types that include these from/to amounts. If we end up needing to make this more localized in the future, we can handle that in a separate ticket.